tests/test_editor_designer.py

```diff
@@ -0,0 +1,730 @@
+            attr.resolveReferences = mock.Mock()
+            attr.json_value = "j" + str(i)
+            if i == 1 or i == 2:
+                attr.turned_on = True
+                attr.mandatory = True
+            else:
+                attr.turned_on = False
+                attr.mandatory = False
+            mock_dict["key" + str(i)] = attr
+            attr.createWidget = mock.Mock(return_value=QtWidgets.QWidget())
+            attr.defaultCopy = mock.Mock(return_value=mock_copies[i - 1])
+
+        mock_attributes = mock.MagicMock()
+        mock_attributes.attributes = mock_dict
+        mock_attributes.__iter__.return_value = mock_attributes.attributes.items()
+        mock_attributes.__getitem__ = mock_dict.__getitem__
+        json_object = ja.DirectlyEditableObject(mock_stack, mock_attributes)
+        self.assertIs(mock_attributes, json_object.value)
+        for attr in mock_dict.values():
+            attr.connectParent.assert_called_with(json_object)
+
+        json_object.resolveReferences()
+        for attr in mock_dict.values():
+            attr.resolveReferences.assert_called()
+
+        self.assertEqual({"key1": "j1", "key2": "j2"}, json_object.json_value)
+        json_object.json_value = {"key1": "n1", "key2": "n2", "key3": "n3"}
+        for i, m in enumerate(mock_dict.values()):
+            self.assertEqual("n" + str(i + 1), m.json_value)
+            self.assertEqual(True, m.turned_on)
+
+        json_object.json_value = {"key2": "m2"}
+        self.assertTrue(mock_dict["key1"].turned_on)
+        self.assertTrue(mock_dict["key2"].turned_on)
+        self.assertFalse(mock_dict["key3"].turned_on)
+
+        panel = json_object.createPanel()
+        self.assertIsInstance(panel, QtWidgets.QWidget)
+        for attr in mock_dict.values():
+            attr.createWidget.assert_called_once()
+
+        for attr in mock_dict.values():
+            attr.createWidget = mock.Mock(return_value=QtWidgets.QWidget())
+        widget = json_object.createEditWidget()
+        self.assertIsInstance(widget, QtWidgets.QWidget)
+        for attr in mock_dict.values():
+            attr.createWidget.assert_called_once()
+
+        copy_object = json_object.defaultCopy()
+        self.assertIsInstance(copy_object, ja.JsonObject)
+        for i, key_attr in enumerate(copy_object.value()):
+            key, attr = key_attr
+            self.assertIsInstance(attr, mock_copies[i])
+
+    def testObjectList(self):
+        mock_stack = mock.Mock()
+        mock_stack.addObject = mock.Mock()
+        mock_object = mock.Mock()
+        mock_object.connectParent = mock.Mock()
+        mock_object.resolveReferences = mock.Mock()
+        mock_object.been_set = TestSignal()
+
+        mock_dict = {}
+        for i in range(3):
+            mock_attr = mock.Mock()
+            mock_attr.been_set = TestSignal()
+            mock_attr.value = "val" + str(i + 1)
+            mock_dict["key" + str(i + 1)] = mock_attr
```

Comment:
use python string formatting instead of plus i.e **f"val {i+1}"**. Please replace every where plus is used to concatenate strings